engines/ic3ia.cpp

```diff
@@ -0,0 +1,369 @@
+  }
+
+  assert(!r.is_unknown());
+  return r.is_sat();
+}
+
+Conjunction IC3IA::get_conjunction_from_model()
+{
+  TermVec conjuncts;
+  conjuncts.reserve(pred_statevars_.size());
+  Term val;
+  for (auto p : pred_statevars_) {
+    if ((val = solver_->get_value(p)) == true_) {
+      conjuncts.push_back(p);
+    } else {
+      assert(val == false_);
+      conjuncts.push_back(solver_->make_term(Not, p));
+    }
+  }
+  return Conjunction(solver_, conjuncts);
+}
+
+void IC3IA::set_labels()
+{
+  assert(solver_context_ == 0);  // expecting to be at base context level
+  // set semantics of labels
+  Sort boolsort = solver_->make_sort(BOOL);
+  if (!init_label_) {
+    // frame 0 label is identical to init label
+    init_label_ = frame_labels_[0];
+  }
+  if (!trans_label_) {
+    trans_label_ = solver_->make_symbol("__trans_label", boolsort);
+    // Note: Using iff instead of implies so that we can negate it
+    // for example in relational preimage generalization
+    solver_->assert_formula(
+        solver_->make_term(Iff, trans_label_, abs_ts_.trans()));
+
+    // add boolean state variables as predicates
+    // NOTE: this must be done before adding predicates
+    // because adding predicates also creates a new state var
+    // for each predicate and we don't want to dobule count those
+    for (auto sv : abs_ts_.statevars()) {
+      if (boolsort_ == sv->get_sort()) {
+        pred_statevars_.push_back(sv);
+      }
+    }
+
+    // add all the predicates from init and property
+    UnorderedTermSet preds;
+    get_predicates(ts_.init(), boolsort_, preds, false);
+    get_predicates(bad_, boolsort_, preds, false);
+    for (auto p : preds) {
+      add_predicate(p);
+    }
+  }
+}
+
+bool IC3IA::only_curr(Term & t) { return abs_ts_.only_curr(t); }
+
+Term IC3IA::next(const Term & t) const { return abs_ts_.next(t); }
+
+void IC3IA::set_invar(size_t i)
+{
+  Term Fi = get_frame(i);
+  // need to translate to actual predicates
+  invar_ = solver_->substitute(Fi, predvar2pred_);
+}
+
+void IC3IA::add_predicate(const Term & pred)
+{
+  assert(abs_ts_.only_curr(pred));
+  // add predicate to abstraction and get the new constraint
+  Term predabs_rel = ia_.add_predicate(pred);
+  // refine the transition relation incrementally
+  // by adding a new constraint
+  assert(!solver_context_);  // should be at context 0
+  solver_->assert_formula(
+      solver_->make_term(Implies, trans_label_, predabs_rel));
+
+  // create a new state variable for the IC3 procedure
+  // used to represent the predicate
+  // NOTE: don't need to actually add to the TS
+  //       kept separate for predicate abstraction
+  // TODO: add infrastructure for getting a fresh symbol -- might fail if name
+  // already exists
+  string name = "__pred" + std::to_string(pred_statevars_.size());
+  Term pred_state = abs_ts_.make_statevar(name, boolsort_);
+  pred_statevars_.push_back(pred_state);
+  predvar2pred_[pred_state] = pred;
```

Comment:
why do we need to introduce new boolean predicate variables?

Reply:
To be honest, I'm just basing it entirely off the available open-source implementation which introduces indicator variables like that. I don't see any reason why it wouldn't work if we just use the predicates themselves, so I can make that change. It would be simpler.

Reply:
We decided offline to not use the indicator variables here, made the change with a1501f856fa66adab0e293b53f0ea121a22168a1.